Add globals support

101arrowz opened the pull request:
<!---
Thanks for filing a pull request 😄 ! Before you submit, please read the following:

Search open/closed issues before submitting since someone might have pushed the same thing before!
-->

# ↪️ Pull Request

<!---
Provide a general summary of the pull request here
Please look for any issues that this PR resolves and tag them in the PR.
-->

Adds support for globals using the `alias` in `package.json`.

Based off of #4846 and #4850.

Fixes #3305.

## 💻 Examples

<!-- Examples help us understand the requested feature better -->
In `package.json`:
```json
{
    "alias": {
        "react": { "global": "React" },
        "myAliasedPackageName": { "fileName": "./myPkg.js" },
        "finalPackage": "finalPackage.js"
    }
}
```
All of the above syntax forms work, for globals and normal aliases respectively. <strike>The pure string syntax works less reliably because it has to guess that it's not an alias based on whether or not the file exists.</strike> (Removed)


## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [x] Included links to related issues/PRs

mischnic commented:
Why do we need so many variants of specifying a global? Does some other tool have this?

I think just adding `{"global": "React"}` is less confusing (and not ambigious)
```json
{
    "alias": {
        "react": { "global": "React" },
        "lodash": "always-a-filename-alias.js"
    }
}
```

101arrowz commented:
I personally like multiple ways of specifying options. For sure, I think having it possible to do just
```json
{
     "jquery": "$"
}
```
would be important because developers would find that more intuitive and less clunky, especially since they're very unlikely to have a file named `$`. I could remove the `["global", "$"]` method (although right now it's just a byproduct of how I'm representing aliases and could just remain undocumented).

If you'd like to make it a rigid structure, I can do that for sure, but it wouldn't be as user-friendly.

mischnic commented:
> would be important because developers would find that more intuitive and less clunky,

macOS has a case-insensitive filesystem, so `"react": "React"` wouldn't even be stable across OSs
(if it would check `if(does package of that name exist){ usePackage()} else { useGlobal() }`)

101arrowz commented:
Fair point, but they could always just use `window.React` if absolutely necessary. I'll remove the string syntax if you insist, though.

101arrowz commented:
Any feedback on the implementation?

mischnic commented on packages/core/integration-tests/test/globals.js:
```diff
@@ -0,0 +1,18 @@
+      'utf8',
+    );
+
+    assert(/module\.exports\ =\ React/.test(index));
```
Testing this using  `module.exports = React.createElement...` and
```js
let output = await run(b, {React: {createElement(tag){ return "ok"; }}});

assert.equals(output, "ok");
```
would be less brittle.

101arrowz replied:
Fixed in 825a3eace291139a66f74e3df8d5a072488fbb97.

101arrowz commented:
@mischnic If there's nothing else that needs changes, I'd recommend merging before the next beta. I have a couple projects I'd be able to move from Webpack to Parcel 2 if Parcel 2 supported globals. By the way, the reason I recommended string syntax was for [similarity to Webpack](https://webpack.js.org/configuration/externals/).

mischnic commented:
The tests are failing

> I think just adding {"global": "React"} is less confusing (and not ambigious)

We still think this is the right path to take (only having the object and not an array). If someone reads the docs for the first time, the first question would probably be: "why are there two ways to achieve the same thing? There has to be *something*"

> recommended string syntax

Makes sense, but that does unfortunately not really work with `alias`

101arrowz commented:
I have already removed the string syntax (check the commit history). I have now fixed the failing tests (at least on my machine), thanks for pointing it out.

101arrowz commented:
I have already removed the string syntax (check the commit history). I have now fixed the failing tests (at least on my machine), thanks for pointing it out.
